pkg/admin/migration/migration.go

```diff
@@ -0,0 +1,282 @@
+			defer wg.Done()
+			objectList := &unstructured.UnstructuredList{}
+			var migrationErr error
+
+			migrationErr = retry.OnError(retry.DefaultBackoff, canRetry, func() error {
+				objectList, migrationErr = d.list(ctx, sch, metav1.ListOptions{})
+				if migrationErr != nil {
+					return migrationErr
+				}
+				return nil
+			})
+
+			if migrationErr != nil {
+				errorOccured = true
+				migrationErr = fmt.Errorf("could not list resources: %+v", migrationErr)
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{Error: migrationErr, ResourceVersion: sch, Timestamp: time.Now()})
+				lock.Unlock()
+				return
+			}
+
+			for _, object := range objectList.Items {
+				ref := object
+				migrationErr := d.migrateOneItem(ctx, sch, &ref)
+				if migrationErr != nil {
+					errorOccured = true
+				}
+				lock.Lock()
+				results.Items = append(results.Items, MigrationResult{
+					Error:           migrationErr,
+					ResourceVersion: sch,
+					ObjectMeta:      getObjectMeta(&ref),
+					Timestamp:       time.Now()})
+				lock.Unlock()
+			}
+		}(&wg, sch)
+	}
+	if errorOccured {
+		results.Status = MigrationFailure
+	}
+	wg.Wait()
+	klog.Infof("schema migration finished, it took %s to complete", time.Since(start).String())
+	return results, nil
+}
+
+func (d *migrator) findMigratableResources(ctx context.Context) ([]schema.GroupVersionResource, error) {
+	customGroups, err := d.findCustomGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+	aggregatedGroups, err := d.findAggregatedGroups(ctx)
+	if err != nil {
+		return nil, err
+	}
+	resourceToGroupVersions := make(map[string][]schema.GroupVersion)
+	_, resourceLists, err := d.discoveryClient.ServerGroupsAndResources()
+	if err != nil {
+		return nil, err
+	}
+	for _, resourceList := range resourceLists {
+		gv, err := schema.ParseGroupVersion(resourceList.GroupVersion)
+		if err != nil {
+			klog.Errorf("cannot parse group version %s, ignored", resourceList.GroupVersion)
+			continue
+		}
+		if customGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's a custom group", gv.Group)
+			continue
+		}
+		if aggregatedGroups.Has(gv.Group) {
+			klog.V(4).Infof("ignored group %v because it's an aggregated group", gv.Group)
+			continue
+		}
+		for _, r := range resourceList.APIResources {
+			// ignore subresources
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			if blackListResources.Has(r.Name) {
+				continue
+			}
+			// ignore resources that cannot be listed and updated
+			if !sets.NewString(r.Verbs...).HasAll("list", "update") {
+				continue
+			}
+			gvs := resourceToGroupVersions[r.Name]
+			gvs = append(gvs, gv)
+			resourceToGroupVersions[r.Name] = gvs
+		}
+	}
+
+	ret := []schema.GroupVersionResource{}
+	for resource, groupVersions := range resourceToGroupVersions {
+		if len(groupVersions) == 1 {
+			continue
+		}
+		ret = append(ret, groupVersions[0].WithResource(resource))
```

Comment:
So what happens here: we are building a list of GVRs that at least two `groupVersion`s?
We don't care about `len == 0` because if there's `resource : groupVersions` in this map (added in L178-180), then `groupVersions` has at least 1 item?

Some explanation in the code would be nice

Reply:
Good point, let me add some comments there and link back to their original documentation as well. One note, the original docs call out a TODO for improvements, but I can't find any indication that, that KEP has been implemented.